### ISSUE TITLE
Sending arguments to scripts when using DDP for multi-GPU training

### ISSUE THREAD

**Issue by finlaymorrison** [ultralytics/ultralytics#5989]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Currently there is no way to pass command-line arguments to the processes started for DDP, leading to errors when your scripts require arguments. This can be fixed by adding an additional argument to the train function which allows you to specify list of command line arguments to append to the end of the call to `torch.distributed.launch`.

### Use case

Running scripts which require command-line arguments when using DDP for multi-GPU training.

### Additional

There is a potential workaround by setting environment variables in the script the first time it is run, and then checking the environment variables in subsequent runs by the worker processes. This is rather clunky and so it would be great for this to be implemented in the train function. I already have some code working for this, so I would be happy to submit a PR to get this sorted.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@finlaymorrison hello, and thank you for your suggestion and offer to contribute to the improvement of YOLOv8's multi-GPU training capabilities. 

Your idea for the enhanced train function to accommodate command-line arguments for Distributed Data Parallel (DDP) processes seems like a valuable addition. We always strive to create a more flexible and efficient user experience, especially for complex procedures such as multi-GPU training.

The workaround you've come up with using environment variables is certainly creative, but as you said, it might lead to a clunkier setup. Therefore, directly integrating command-line arguments into the train function could indeed be more streamlined.

We're looking forward to reviewing your proposed changes. When you're ready, please feel free to submit a pull request, including an explanation of your modifications and any associated code. Before submission, ensure your code aligns with our contribution guidelines and standards to maintain the harmony of the code.

This is how Open Source grows and we appreciate your initiative!

Best,
Glenn

**Comment by finlaymorrison** [comment]
Hi Glen,

I have already sorted out a fix to this in PR #5990, is there anything else that I need to do?

**Comment by glenn-jocher** [comment]
@finlaymorrison hello,

Thank you for your contribution. If you have already created a pull request, the next step would be to make sure your changes pass all the tests. We will also review your code to ensure it aligns with our project standards and does not introduce any unexpected behavior. 

Please ensure your pull request includes a clear and concise description of the changes made, along with any additional context that could be helpful in understanding your solution.

If there are any concerns or required changes during our review, we will communicate them on the PR page.

Thank you again for your contribution to YOLOv8. We look forward to reviewing your code.

Best,
Glenn

**Comment by glenn-jocher** [comment]
@finlaymorrison upon review I don't think any code changes are required. Have you reviewed https://docs.ultralytics.com/modes/train/#multi-gpu-training?

The command line arguments for DDP are passed just like any other training commands, i.e.:
```bash
yolo detect train data=coco128.yaml model=yolov8n.pt epochs=100 imgsz=640 device=0,1
```

Or for Python:

```python
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8n.pt')  # load a pretrained model (recommended for training)

# Train the model with 2 GPUs
results = model.train(data='coco128.yaml', epochs=100, imgsz=640, device=[0, 1])
```

**Comment by finlaymorrison** [comment]
I'm not sure what you are trying to show with that code, as I do not think that has anything to do with my use case. The script I am
using can be seen below.

```python
from ultralytics import YOLO
import argparse

os.environ["KMP_DUPLICATE_LIB_OK"]="TRUE"

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('-d', '--data', type=str, default=None, help='Path to dataset.')
    parser.add_argument('-m', '--model', type=str, default=None, help='YOLO Model to train.')
    parser.add_argument('-o', '--output', type=str, default=None, help='Path to output directory.')
    parser.add_argument('-s', '--stage', type=str, default=None, help='Model variant to train.')
    args = parser.parse_args()

    cfg = params_show(stages=args.stage)[args.stage]
    devices = [i for i in range(torch.cuda.device_count())]

    model = YOLO(f"{args.model}{cfg['variant']}.yaml")

    model.train(data=os.path.abspath(args.data), epochs=cfg["epochs"], batch=cfg["batch"], project=args.output, single_cls=True, weight_decay=0.0007, imgsz=793, device=devices)


if __name__ == "__main__":
    main()
```

The issue is that when I run the script with the 4 arguments 'data', 'model', 'output', 'stage', the multi-GPU is implemented under the hood by running the script again with a process for each of the GPUs. However, when these processes are started those 4 command line arguments are not passed in, and so I get an error from argparse as there are missing required arguments.

**Comment by finlaymorrison** [comment]
To add onto this, with the changes made by the PR, you can fix this issue by using the following code to train the model:
```python
dist_args = ["--data", args.data, "--model", args.model, "--output", args.output, "--stage", args.stage]
model.train(data=os.path.abspath(args.data), epochs=cfg["epochs"], batch=cfg["batch"], project=args.output, single_cls=True, weight_decay=0.0007, imgsz=793, device=devices, distributed_args=dist_args)
```

**Comment by glenn-jocher** [comment]
@finlaymorrison got it, thanks for the code to reproduce. I've worked on a fix in #6009 that appears to resolve the issue, allowing for DDP training via Python, `yolo` CLI commands, and custom Python files with argparse:

```python
import argparse

from ultralytics import YOLO


def main(epochs: int, imgsz: int, device: str):
    """
    Train a YOLO model with the given epochs and image size.

    Args:
        epochs (int): Number of epochs for training.
        imgsz (int): Image size for training.
        device (str): Device for training.
    """
    # Load a model
    model = YOLO('yolov8n.pt')  # Load a pretrained model (recommended for training)

    # Train the model
    results = model.train(data='coco128.yaml', epochs=epochs, imgsz=imgsz, device=device)
    print(f"Training complete. Results: {results}")


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Train a YOLO model.')
    parser.add_argument('--epochs', type=int, default=100, help='Number of epochs for training.')
    parser.add_argument('--imgsz', type=int, default=640, help='Image size for training.')
    parser.add_argument('--device', type=str, default='', help='Device for training.')
    args = parser.parse_args()

    main(args.epochs, args.imgsz, args.device)

# Run 'python example.py --epochs 3 imgsz 320 device=0,1,2,3'
```

### Output

<img width="1788" alt="Screenshot 2023-10-29 at 20 05 50" src="https://github.com/ultralytics/ultralytics/assets/26833433/d905aa65-f423-4f61-b7ef-a19b83fcd000">

**Comment by glenn-jocher** [comment]
@finlaymorrison great news 😃! Your original issue may now be resolved ✅ in PR #6009 due for publication in `ultralytics==8.0.203` shortly. 

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!

**Comment by finlaymorrison** [comment]
Just checked your solution out --- works perfectly. Thanks!

**Comment by glenn-jocher** [comment]
@finlaymorrison that's great to hear! I'm glad the solution is working well for you. Don't hesitate to reach out if you come across any other issues or if there's anything else we can assist you with. Thank you for contributing to YOLOv8, your feedback is invaluable to us!
